### PR TITLE
Scheduler rework -- part two, work items

### PR DESCRIPTION
**Pull request by vpodzime**
Follow-up to #104

### CODE REVIEW

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 1]
```diff
@@ -137,24 +126,20 @@ mender_scheduler_init(void) {
...
+    assert(NULL != work);
 
     /* Create work context */
-    memset(work_context, 0, sizeof(mender_scheduler_work_context_t));
+    mender_platform_work_t *work_context = (mender_platform_work_t *)malloc(sizeof(mender_platform_work_t));
+    if (NULL == work_context) {
+        mender_log_error("Unable to allocate memory");
+        goto FAIL;
+    }
+    memset(work_context, 0, sizeof(mender_platform_work_t));
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:142]: Can consider using `calloc()` 
```suggestion
    mender_platform_work_t *work_context = (mender_platform_work_t *)calloc(1, sizeof(mender_platform_work_t));
    if (NULL == work_context) {
        mender_log_error("Unable to allocate memory");
        goto FAIL;
    }
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 2]
```diff
@@ -181,91 +166,140 @@ mender_scheduler_work_create(mender_scheduler_work_params_t *work_params, mender
...
 FAIL:
 
-    mender_scheduler_work_delete(work_context);
+    /* Release memory */
+    if (NULL != work_context) {
+        timer_delete(work_context->timer_handle);
+        pthread_mutex_destroy(&work_context->sem_handle);
+        if (NULL != work_context->params.name) {
+            free(work_context->params.name);
+        }
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:182]:
```suggestion
        free(work_context->params.name);
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 3]
```diff
@@ -181,91 +166,140 @@ mender_scheduler_work_create(mender_scheduler_work_params_t *work_params, mender
...
+}
+
+mender_err_t
+mender_scheduler_work_set_period(mender_work_t *work, uint32_t period) {
+    assert(NULL != work);
+
+    /* Set timer period */
+    work->params.period = period;
+    struct itimerspec its;
+    memset(&its, 0, sizeof(struct itimerspec));
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:231]:
```suggestion
    struct itimerspec its = {0};
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 4]
```diff
@@ -181,91 +166,140 @@ mender_scheduler_work_create(mender_scheduler_work_params_t *work_params, mender
...
+mender_scheduler_work_deactivate(mender_work_t *work) {
+    assert(NULL != work);
+
     /* Check if the work was activated */
-    if (work_context->activated) {
+    if (work->activated) {
 
         /* Stop the timer used to periodically execute the work (if it is running) */
         struct itimerspec its;
         memset(&its, 0, sizeof(struct itimerspec));
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:265]:
```suggestion
        struct itimerspec its = {0};
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 5]
```diff
@@ -181,91 +166,140 @@ mender_scheduler_work_create(mender_scheduler_work_params_t *work_params, mender
...
-    mender_scheduler_work_delete(&main_work_context);
+mender_scheduler_work_delete(mender_work_t *work) {
+    assert(NULL != work);
 
+    /* Release memory */
+    timer_delete(work->timer_handle);
+    pthread_mutex_destroy(&work->sem_handle);
+    if (NULL != work->params.name) {
+        free(work->params.name);
+    }
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:293]:
```suggestion
    free(work->params.name);
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 6]
```diff
@@ -276,6 +310,63 @@ mender_scheduler_exit(void) {
...
 
+static void
+mender_scheduler_timer_callback(union sigval timer_data) {
+    /* Get work context */
+    mender_platform_work_t *work = (mender_platform_work_t *)timer_data.sival_ptr;
+    assert(NULL != work);
+
+    /* Exit if the work is already pending or executing */
+    struct timespec timeout;
+    memset(&timeout, 0, sizeof(struct timespec));
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:321]:
```suggestion
    struct timespec timeout = {0};
```

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 7]
```diff
@@ -276,6 +310,63 @@ mender_scheduler_exit(void) {
...
+        if (NULL == work) {
+            goto END;
+        }
+
+        /* Call work function */
+        if (MENDER_DONE == work->params.function()) {
+
+            /* Work is done, stop timer used to execute the work periodically */
+            struct itimerspec its;
+            memset(&its, 0, sizeof(struct itimerspec));
```
Comment (larsewi) [inline platform/scheduler/posix/src/mender-scheduler.c:351]:
```suggestion
            struct itimerspec its = {0};
```

File: platform/scheduler/zephyr/src/mender-scheduler.c [hunk 8]
```diff
@@ -42,87 +42,208 @@
...
+    assert(NULL != work_params->name);
+    assert(NULL != work);
+
+    /* Create work context */
+    mender_platform_work_t *work_context = malloc(sizeof(mender_platform_work_t));
+    if (NULL == work_context) {
+        mender_log_error("Unable to allocate memory");
+        goto FAIL;
+    }
+    memset(work_context, 0, sizeof(mender_platform_work_t));
```
Comment (larsewi) [inline platform/scheduler/zephyr/src/mender-scheduler.c:98]:
```suggestion
    mender_platform_work_t *work_context = calloc(1, sizeof(mender_platform_work_t));
    if (NULL == work_context) {
        mender_log_error("Unable to allocate memory");
        goto FAIL;
    }
```

File: platform/scheduler/zephyr/src/mender-scheduler.c [hunk 9]
```diff
@@ -42,87 +42,208 @@
...
+}
+
+mender_err_t
+mender_scheduler_work_delete(mender_work_t *work) {
+    assert(NULL != work);
+
+    /* Release memory */
+    if (NULL != work->params.name) {
+        free(work->params.name);
+    }
```
Comment (larsewi) [inline platform/scheduler/zephyr/src/mender-scheduler.c:200]:
```suggestion
    free(work->params.name);
```

**Review by larsewi — APPROVED** [review]
Looks good 🚀 I know you've mainly been copying code around, so feel free to dismiss nitpicks.

File: platform/scheduler/posix/src/mender-scheduler.c (hunk 1 above)
Comment (vpodzime) [reply platform/scheduler/posix/src/mender-scheduler.c:142]: Good point!

**Review by danielskinstad — APPROVED** [review]
Looks good to me

**Pull request merged**